Add `.get_name()` to `Function_Stmt` and `Entity_Decl`

ZedThree opened the pull request:
Fixes #225

Probably I've not added the tests to the best location, so I'm happy to move them if there's somewhere better

arporter commented:
Many thanks for this enhancement. Note that one of the conditions of code review (https://github.com/stfc/fparser/wiki/CodeReview) is that there is full test coverage of any modified code. The tests themselves need to be in files named for the classes/rules to which they apply. Therefore, the one for Function_Stmt is good but the other needs to be moved. If there are is no such file yet (and for many classes there aren't) then please copy one of the existing ones and add your tests. Please also search the massive fortran2003_test.py and move any tests of that class into the new file too.

ZedThree commented:
Done! I have also added tests for the `get_name` methods on `Program_Stmt` and `Subroutine_Stmt`

ZedThree commented:
I notice that some classes use `property` for some of the `items`, e.g. `label` of `Do_Label_Stmt`. That would probably be a nicer interface for `name` here too.

Probably something for a different PR, but is there any interest in converting all the accessor methods for the various `items` elements to properties instead? e.g. `get_name()` -> `name`?

arporter commented on src/fparser/two/Fortran2003.py:
```diff
@@ -2935,6 +2935,13 @@ def tostr(self):
+    def get_name(self):
+        '''Provides the entity name as an instance of the `Name` class.
+
+        :rtype: `Name`
```
Please could you use Sphinx auto-api markup for this type, i.e. ":py:class:`fparser.two.Fortran2003.Name`" (without the double quotes).

ZedThree replied:
Sure, will do! [`default_role`](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-default_role) might be a useful config for Sphinx

arporter commented on src/fparser/two/Fortran2003.py:
```diff
@@ -10570,6 +10577,13 @@ def tostr(self):
+    def get_name(self):
+        '''Provides the function name as an instance of the `Name` class.
+
+        :rtype: `Name`
```
Please could you use Sphinx markup again here.

arporter commented on src/fparser/two/tests/fortran2003/test_function_stmt_r1224.py:
```diff
@@ -61,3 +61,8 @@ def test_function_new_symbol_table(f2003_create):
+
+
+def test_function_get_name():
+    obj = Function_Stmt("function foo()")
```
Please could you add a (very) brief docstring saying what this test is for.

arporter commented on src/fparser/two/tests/fortran2003/test_module_r1104.py:
```diff
@@ -56,3 +56,8 @@ def test_module_new_symbol_table(f2003_create):
+
+
+def test_module_get_name():
+    obj = Module_Stmt("module foo")
```
Please add a brief docstring for this test.

arporter commented on src/fparser/two/tests/fortran2003/test_program_stmt_r1102.py:
```diff
@@ -75,3 +75,13 @@ def test_prog_symbol_table():
+
+
+def test_get_name():
+    obj = Program_Stmt("program foo")
```
Please add brief docstrings for these new tests.

arporter commented on src/fparser/two/tests/fortran2003/test_program_stmt_r1102.py:
```diff
@@ -75,3 +75,13 @@ def test_prog_symbol_table():
     assert "my_prog" in SYMBOL_TABLES._symbol_tables
+
+
+def test_get_name():
```
Please mark the two new tests with `@pytest.mark.usefixtures("f2003_create")` as otherwise they don't work in isolation and may fail when run in parallel (e.g. `pytest fortran2003/test_program_stmt_r1102.py::test_get_name` fails currently).

ZedThree replied:
Is there a difference between using the decorator and passing the fixture as an argument? Is it just that the return value of `f2003_create` isn't needed in the tests?

arporter replied:
Yes, the return value isn't needed and pylint will complain about an unused variable.

arporter commented on src/fparser/two/tests/fortran2003/test_module_r1104.py:
```diff
@@ -56,3 +56,8 @@ def test_module_new_symbol_table(f2003_create):
     assert "a" in SYMBOL_TABLES._symbol_tables
+
+
+def test_module_get_name():
```
Please mark this test with `@pytest.mark.usefixtures("f2003_create")`

arporter commented on src/fparser/two/tests/fortran2003/test_function_stmt_r1224.py:
```diff
@@ -61,3 +61,8 @@ def test_function_new_symbol_table(f2003_create):
     assert "a" in SYMBOL_TABLES._symbol_tables
+
+
+def test_function_get_name():
```
Please mark this test with `@pytest.mark.usefixtures("f2003_create")`

arporter commented on src/fparser/two/tests/fortran2003/test_subroutine_stmt_r1232.py:
```diff
@@ -56,3 +56,8 @@ def test_sub_stmt_new_symbol_table(f2003_create):
     assert "a" in SYMBOL_TABLES._symbol_tables
+
+
+def test_subroutine_get_name():
```
Please add a `@pytest.mark.usefixtures("f2003_create")`

arporter commented on src/fparser/two/tests/test_entity_decl.py:
```diff
@@ -0,0 +1,83 @@
+# Modified work Copyright (c) 2017-2022 Science and Technology
```
Many thanks for adding this file. I have to confess that when I first looked at this I was assuming you were a new contributor from the Met Office otherwise I wouldn't have been quite so stringent in my requests!

arporter replied:
Please could you move this file into the `tests/fortran2003` directory.

arporter commented on src/fparser/two/tests/test_entity_decl.py:
```diff
@@ -0,0 +1,83 @@
+import six
+
+
+def test_entity_decl_repr():
```
Please add a `@pytest.mark.usefixtures("f2003_create")` and a docstring for these tests. In fact, if there's a simple way to add autouse for "f2003_create" then feel free.

arporter commented on src/fparser/two/tests/test_fortran2003.py:
```diff
@@ -1029,42 +1029,6 @@ def test_intent_attr_spec():  # R503.f
     assert _repr_utf(obj) == "Intent_Attr_Spec('INTENT', Intent_Spec('IN'))"
 
 
```
Great to shrink this file a bit - thanks very much.

arporter reviewed (changes requested):
Great work, thanks for doing all of that.
You'll see I've requested (brief) docstrings for the tests and they also need the `f2003_parser` fixture setup. Currently we seem to do that on a test-by-test basis so I'm happy for you to copy that although ultimately we really need to add an autouse for the files where it makes sense.
I've also updated the PR title so that #225 will get closed when this is merged.

ZedThree commented:
All done! @arporter ready for review

I've added `autouse=True` to the `f2003_create` and `f2008_create` fixtures, so that all tests in the `f2003`/`f2008` subdirectories use them automatically.

> I've also updated the PR title so that #225 will get closed when this is merged.

Just to note that the `Fixes #225` in the PR comment body also does this :slightly_smiling_face:

arporter commented on src/fparser/two/Fortran2003.py:
```diff
@@ -9244,8 +9244,8 @@ class containing the name of the program.
 
-        :returns: the program name as a `Name` class
-        :rtype: `Name`
+        :returns: the program name as a :py:class:`Name` class
```
There's no need to duplicate the type specification in the :returns: field but it's not a biggie.

arporter reviewed (approved):
All requested changes have been made (thanks).
Tests now pass in isolation.
No documentation update required.
Issue number not mentioned anywhere.
I will proceed to merge this.
